### DISCUSSION TITLE
Can I get the form instance index (within a collection) in the form class?

### DISCUSSION THREAD

**Discussion by lsmoker** [jrief/django-formset#121] [Q&A] [answered]
I would like my `hide_condition` on a form to be dynamic based on the form index within the collection that the form is nested in. Can this be done? I'm doing this so my collapse toggle displays for each Opportunities collection which is nested in a Bundles collection. Finding that `'bundles.0.collapse_opp.toggle'` works for the first instance was a first step.

forms.py
```
class CollapseForm(forms.Form):
    toggle = fields.BooleanField(
        required=False,
    )

    def __init__(self, *args, **kwargs):
        self.collapse_label = kwargs.pop('collapse_label')
        super().__init__(*args, **kwargs)
        self.fields['toggle'].label = self.collapse_label

        
class OpportunityForm(FieldsetMixin, forms.ModelForm):
    # this works as long as there's one bundle per credit report
    # the value is the id of the DOM field without 'id_'
    # the 0 should really be the form index in the collection to be dynamic
    hide_condition = 'bundles.0.collapse_opp.toggle'

    id = fields.IntegerField(widget=widgets.HiddenInput)

    class Meta:
        model = Opportunity
        fields = ['id', 'status', 'name', 'description', 'payment', 'rate']


class OpportunityCollection(FormCollection):
    min_siblings = 0
    related_field = 'bundle'

    opportunity = OpportunityForm()


class BundleForm(forms.ModelForm):
    id = fields.IntegerField(widget=widgets.HiddenInput)

    class Meta:
        model = Bundle
        fields = ['id', 'name', 'type', 'status']


class BundleCollection(FormCollection):
    min_siblings = 0
    related_field = 'report'

    bundle = BundleForm()
    collapse_opp = CollapseForm(collapse_label="Hide Opportunities")
    opportunities = OpportunityCollection()


class CreditReportCollection(FormCollection):
    credit_report = CreditReportForm()
    bundles = BundleCollection()
```

views.py
```
class CreditReportCollectionView(EditCollectionView):
    model = CreditReport
    collection_class = CreditReportCollection
    success_url = '/'
    template_name = "bundle_tool_app/view_bundle_opportunites.html"
    extra_context = {
        'click_actions': 'disable -> submit -> reload !~ scrollToError',
        "title": "Manage bundled opportunities",
        'force_submission': False,
    }
```

**Comment by jrief** [comment]
Hmm, difficult. The problem here is that it is hard to know how often a FormCollection with siblings is instantiated. If I understand correctly, you want a hide_condition such as `'bundles.13.collapse_opp.toggle'` (here 13 is just a number) or how would you specify that?

**Comment by lsmoker** [comment]
Ideally, I would specify `show_condition = 'bundles.${siblingId}.collapse_opp.toggle'` and then that token would be dynamically substituted with the sibling id from that instance of my OpportunityForm within the collection. Or can that property be set within a form method that can introspect it's own instance?

**Reply by jrief** [reply]
That might be possible. I'll check for it. Will not happen before version 1.5.

One problem here is that there might many levels of `siblingId`-s since you can nest form collections with siblings into each other. So we would need something like `siblingId_0`, `siblingId_1`, `siblingId_9`, etc.

Just a question, would it be possible to merge the forms `CollapseForm` and `OpportunityForm` and add a Fieldset below containing 'status', 'name', 'description', 'payment', 'rate'. Then you could use the relative path in your `hide_condition`.

**Comment by lsmoker** [comment]
That would work but then the user be forced to collapse each OpportunityForm instead of all in the collection under the parent Bundle.

And in rethinking my request for the notation of `show_condition = 'bundles.${siblingId}.collapse_opp.toggle'`... what if it was instead `show_condition = 'bundles.*.collapse_opp.toggle'` where the '*' is a wildcard for matching that pattern. Then the form instance id is not even an issue.

**Reply by jrief** [reply]
With a wildcard, you might have some conditions evaluating to true and some to false. It would be quite difficult to find a consistent Boolean expression.

Anyway, thanks for your thoughts and requirements and I'm happy to hear that this feature is used by others as well.

**Comment by lsmoker** [comment]
I guess in my case, I'm looking for a way to show/hide the whole collection instead of one form.

**Reply by lsmoker** [reply]
If show_condition/hide_condition were supported on the collection class, that would solve the problem.

**Comment by lsmoker** [comment, marked as answer]
Here's how I got the sibling ID in the OpportunityForm and then updated the show_condition for each Bundle sibling:
```
    # the "X" will be replaced below (in get_context) with the bundle sibling id
    # within the CreditReportCollection
    show_condition = 'bundles.X.collapse_opp.toggle'

    def get_context(self):
        context = super().get_context()
        # see show_condition assignment above
        bundle_id = self.form_id.split('.')[1]
        context['show_condition'] = context['show_condition'].replace("X", bundle_id)
        return context
```
